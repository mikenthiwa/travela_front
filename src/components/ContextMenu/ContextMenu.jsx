import React, {Component} from 'react';
import PropTypes from 'prop-types';

export default class ContextMenu extends Component {
    state = {
      open: false
    };

    menu = React.createRef();

    componentDidMount() {
      document.addEventListener('mousedown', this.onClickOutside);
    }

    componentWillUnmount() {
      document.removeEventListener('mousedown', this.onClickOutside);
    }

    toggleMenu = (e) => {
      e.preventDefault();
      this.setState((prevState) => ({open: !prevState.open}));
    };

    onClickOutside = (e) => {
      if (!this.menu.current.contains(e.target)) {
        this.setState({open: false});
      }
    };

    onItemClick = (e,onClick) => {
      this.toggleMenu(e);
      onClick(e);
    };

    render() {
      const {open} = this.state;
      const { children , classNames } = this.props;
      return (
        <span>
          <i
            className="fa fa-ellipsis-v on"
            role="presentation"
            onClick={this.toggleMenu}
          />
          <div
            ref={this.menu}
            className={`context-menu ${classNames} ${open && 'open'}`}
          >
            <ul className="table__menu-list">
              {
                (children instanceof Array ? children : [children]).map((child) => {
                  const onClick = child.props.onClick;
                  return React.cloneElement(child, {
                    onClick: (e) => this.onItemClick(e, onClick),
                    key: Math.random()
                  });
                })
              }
            </ul>
          </div>
        </span>
      );
    }
}

ContextMenu.defaultProps = {
  classNames: 'table__menu-container',
};

ContextMenu.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.arrayOf(PropTypes.object),
  ]).isRequired,
  classNames: PropTypes.string
};
