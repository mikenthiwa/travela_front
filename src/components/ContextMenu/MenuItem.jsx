import React, {Component} from 'react';
import PropTypes from 'prop-types';

export default class MenuItem extends Component{

  render() {
    const { classNames, onClick, children } = this.props;
    return (
      <li
        className={classNames}
        role="presentation"
        onClick={onClick}>
        <span>{children}</span>
      </li>
    );
  }
}

MenuItem.defaultProps = {
  onClick: () => {}
};

MenuItem.propTypes = {
  classNames: PropTypes.string.isRequired,
  onClick: PropTypes.func,
  children: PropTypes.oneOfType([
    PropTypes.string.isRequired,
    PropTypes.object.isRequired,
  ]
  ).isRequired
};
