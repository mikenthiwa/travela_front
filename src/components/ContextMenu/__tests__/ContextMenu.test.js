import React from 'react';
import ContextMenu from '../ContextMenu';
import MenuItem from '../MenuItem';

const documentEvents = {

};

document.addEventListener = (event, callback) => {
  documentEvents[event] = callback;
};

const event = {
  preventDefault: jest.fn()
};

describe('<ContextMenu /> component', () => {

  let wrapper;
  const editCallback = jest.fn();

  beforeEach(() => {
    wrapper = mount(
      <ContextMenu>
        <MenuItem classNames="edit" onClick={editCallback}>
        Edit
        </MenuItem>
      </ContextMenu>);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should close the menu on click outside', () => {
    documentEvents['mousedown']({ target: document.createElement('p')});

    expect(wrapper.state().open).toBeFalsy();
  });

  it('should open the menu', () => {
    wrapper.find('.fa.on').simulate('click', event);

    expect(wrapper.state().open).toBeTruthy();
  });


  it('should close the menu when the component unmounts', () => {
    document.removeEventListener = jest.fn();
    wrapper.unmount();

    expect(document.removeEventListener).toHaveBeenCalled();
  });

  it('it should invoke the function when the menu item is clicked', () => {
    wrapper.find('li.edit').simulate('click', event);
    expect(editCallback).toHaveBeenCalled();
  });
});
