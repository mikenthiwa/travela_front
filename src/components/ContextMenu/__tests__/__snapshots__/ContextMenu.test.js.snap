// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContextMenu /> component should render correctly 1`] = `
<span>
  <i
    className="fa fa-ellipsis-v on"
    onClick={[Function]}
    role="presentation"
  />
  <div
    className="context-menu table__menu-container false"
  >
    <ul
      className="table__menu-list"
    >
      <li
        className="edit"
        onClick={[Function]}
        role="presentation"
      >
        <span>
          Edit
        </span>
      </li>
    </ul>
  </div>
</span>
`;
